### ISSUE TITLE
Using HitRecord::t before declaring it 

### ISSUE THREAD

**Issue by Krenodeno** [RayTracing/raytracing.github.io#428]
In Chapter 6, at listing 13, `hit_record` definition is
```cpp
struct hit_record {
    vec3 p;
    vec3 normal;
};
```
and then in listing 14, there is reference to a `t` member that do not exist... yet.
```cpp
bool sphere::hit(const ray& r, double t_min, double t_max, hit_record& rec) const {
[...]
            rec.t = temp;
            rec.p = r.at(rec.t);
            rec.normal = (rec.p - center) / radius;
```
`hit_record`'s member `t` is defined later as the time:
`I know that we’ll also want motion blur at some point, so I’ll also add a time input variable.`
Therefore I cannot see any relation to time in the calculations of `temp` listing 18:
```cpp
auto a = r.direction().length_squared();
auto half_b = dot(oc, r.direction());
auto root = sqrt(discriminant);
auto temp = (-half_b - root)/a;
```
If I'm correct, at this point `t` is more like a depth buffer, as to see listing 19.

**Comment by trevordblack** [comment]
This is a bug. Thank you for catching it.

Unfortunately, `rec.t` refers not to the time, but the parameterization of the ray. Namely:

`orig + t*dir`.

`t` needs to be put back into the initial definition of `hit_record`. And we need to consider renaming it since it's overloading with the `time` variable later on

**Comment by trevordblack** [comment]
Just use

```
struct hit_record {
    vec3 p;
    vec3 normal;
    double t;
    bool front_face;

    inline void set_face_normal(const ray& r, const vec3& outward_normal) {
        front_face = dot(r.direction(), outward_normal) < 0;
        normal = front_face ? outward_normal :-outward_normal;
    }
};
```

for `hit_record`, where `t` is the parameterization of the ray (not time). Neglect the comment `I know that we’ll also want motion blur at some point, so I’ll also add a time input variable.` as it was a brain fart on my part.

There is a new PR to fix this issue.

Thank you for bringing it up

**Comment by MacGamerFR** [comment]
Glad I could help !

**Comment by hollasch** [comment]
Resolved in PR #429

**Comment by junkimu** [comment]
t still needs to be added back to listing 18: "The hittable class with time and side"

**Comment by hollasch** [comment]
Thank you for catching this!

**Comment by hollasch** [comment]
It would be good to use a better name, perhaps `ray_parameter`.

**Comment by hollasch** [comment]
See also #418 regarding our nomenclature for time0 and time1. This should be considered as well when making these changes.

**Comment by hollasch** [comment]
Resolved in PR #625
